lib/Infer/Pruning.cpp

```diff
@@ -41,71 +28,112 @@ std::string getUniqueName() {
 bool PruningManager::isInfeasible(souper::Inst *RHS,
                                  unsigned StatsLevel) {
   for (int I = 0; I < InputVals.size(); ++I) {
-    auto C = LHSValues[I];
-    if (C.hasValue()) {
-      auto Val = C.getValue();
-      if (StatsLevel > 2) {
-        llvm::errs() << "  Input:\n";
-        for (auto &&p : InputVals[I]) {
-          if (p.second.hasValue()) {
-            llvm::errs() << "  Var " << p.first->Name << " : "
-                         << p.second.getValue() << "\n";
-          }
+    if (StatsLevel > 2) {
+      llvm::errs() << "  Input:\n";
+      for (auto &&p : InputVals[I]) {
+        if (p.second.hasValue()) {
+          llvm::errs() << "  Var " << p.first->Name << " : "
+                        << p.second.getValue() << "\n";
         }
-        llvm::errs() << "  LHS value = " << Val << "\n";
       }
+    }
 
-      if (!isConcrete(RHS)) {
-        auto CR = findConstantRange(RHS, InputVals[I]);
-        if (StatsLevel > 2)
-          llvm::errs() << "  RHS ConstantRange = " << CR << "\n";
-        if (!CR.contains(Val)) {
-          if (StatsLevel > 2) {
-            llvm::errs() << "  pruned using CR! ";
+    if (LHSHasPhi) {
+      auto LHSCR = LHSConstantRange[I];
+      auto RHSCR = ConstantRangeAnalysis().findConstantRange(RHS, ConcreteInterpreters[I]);
```

Comment:
ok, so if I understand correctly, every time we invoke the known bits or constant range analysis, we're creating a new object of the analysis class in order to store the cache, and then throwing this away?

what we usually do in this case in souper is create a separate context object and use that-- is that not a good approach here?

Reply:
Yes, that's true. The `ConstantRange` and `KnownBitsAnalysis` classes are just a wrapper around previous `findKnownBits`, `findConstantRange` functions to make them more efficient by making use of cache. This is helpful when some instruction has a lot of "uses".

Since all RHSes are different, we don't try to share the cache across different RHS evaluations as that will lead to misevaluations.